include/libdnf/common/class_impl_ptr.hpp

```diff
@@ -0,0 +1,78 @@
+/*
+Copyright Contributors to the libdnf project.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2.1 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef LIBDNF_COMMON_CLASS_IMPL_PTR_HPP
+#define LIBDNF_COMMON_CLASS_IMPL_PTR_HPP
+
+
+namespace libdnf {
+
+template <class T>
+class ClassImplPtr {
+public:
+    /// Constructs a ClassImplPtr that owns nothing.
+    constexpr ClassImplPtr() noexcept = default;
+
+    /// Constructs a ClassImplPtr that takes ownership of `ptr`.
+    explicit constexpr ClassImplPtr(T * ptr) noexcept { this->ptr = ptr; }
+
+    /// Constructs a ClassImplPtr that owns newly created instance value initialized from `src` or
+    /// owns nothing if `src` owns nothing.
+    ClassImplPtr(const ClassImplPtr & src) : ptr(src.ptr ? new T(*src.ptr) : nullptr) {}
+
+    /// Constructs a ClassImplPtr by transferring ownership from `src` to *this and stores the null pointer in `src`.
+    constexpr ClassImplPtr(ClassImplPtr && src) noexcept : ptr(src.ptr) { src.ptr = nullptr; }
+
+    /// Copy assignment operator. Copies the value pointed to by `src`. Not a pointer itself.
+    /// If the destination owns nothing and `src` points to a value, a new instance of the value initialized from
+    /// `src` is created. It does not do anything if both of them own nothing.
+    ClassImplPtr & operator=(const ClassImplPtr & src) {
+        if (ptr != src.ptr) {
+            if (ptr) {
+                *ptr = *src.ptr;
+            } else {
+                ptr = new T(*src.ptr);
+            }
+        }
+        return *this;
+    }
+
+    /// Move assignment operator. Transfers ownership from `src` to *this and stores the null pointer in `src`.
+    ClassImplPtr & operator=(ClassImplPtr && src) noexcept {
+        if (ptr != src.ptr) {
+            delete ptr;
+            ptr = src.ptr;
+            src.ptr = nullptr;
+        }
+        return *this;
+    }
+
+    ~ClassImplPtr() { delete ptr; }
+
+    constexpr T * operator->() noexcept { return ptr; }
+
+    constexpr const T * operator->() const noexcept { return ptr; }
+
+private:
+    T * ptr{nullptr};  // object is owner of the pointer
+};
```

Comment:
I like it. How about naming the class just ImplPtr and also adding a `make_impl()` analogous to `make_unique()`?

Reply:
* Using the name `ImplPtr` is not a problem. Why not? :-)
* Adding `make_impl ()` - I will think about the benefits of this function.